### ISSUE TITLE
LUA Editor Method Tree is LnF v3. Needs upgraded design to LnF v4

### ISSUE THREAD

**Issue by damiensellier** [damiensellier/CtrlrX#42]
LUA Editor Method Tree is LnF v3. Needs upgraded design to LnF v4

![Capture d’écran 2024-12-18 à 08 31 46](https://github.com/user-attachments/assets/bc99302c-a428-48bb-8c3e-e061b3860a07)

**Comment by damiensellier** [comment]
LOOK FOR : CtrlrValueTreeEditorTree

**Comment by damiensellier** [comment]
CtrlrValueTreeEditor.h

class CtrlrValueTreeEditorTree : public TreeView, public LookAndFeel_V4 // Updated v5.6.31


CtrlrLuaMethodEditor.cpp

    methodTree->LookAndFeel_V4::setColour(TreeView::backgroundColourId, LookAndFeel_V4::findColour(TreeView::backgroundColourId));
    methodTree->LookAndFeel_V4::setColour(TreeView::linesColourId, LookAndFeel_V4::findColour(TreeView::linesColourId));

**Comment by damiensellier** [comment]
ITEM PROPERTIES : 

CtrlrLuaMethodEditor.cpp L708

```
const AttributedString CtrlrLuaMethodEditor::getDisplayString(const ValueTree &item)	const
{
	AttributedString str;
...
```

**Comment by damiensellier** [comment]
CtrlrValueTreeEditor.cpp has treeViewItem definitions

**Comment by damiensellier** [comment]
CtrlrPanelWindowManager.cpp 

Child windows for MIDISettings, LuaEditor etc are launched from there : 

```
CtrlrChildWindow *CtrlrPanelWindowManager::createWindow(const CtrlrPanelWindowManager::WindowType window)
{
	if (isCreated(window))
	{
		return (nullptr);
	}

	CtrlrChildWindow *w = new CtrlrChildWindow (*this);

	switch (window)
	{
		case CtrlrPanelWindowManager::LuaConsole:
			w->setContent (new CtrlrLuaConsole (owner));
			break;

		case CtrlrPanelWindowManager::LuaMethodEditor:
			w->setContent (new CtrlrLuaMethodEditor (owner));
			break;

		case CtrlrPanelWindowManager::ModulatorList:
			w->setContent (new CtrlrPanelModulatorList(owner));
			break;

		case CtrlrPanelWindowManager::LayerEditor:
			w->setContent (new CtrlrPanelLayerList(owner));
			break;

		case CtrlrPanelWindowManager::MIDISettings:
			w->setContent (new CtrlrMIDISettingsDialog(owner));
	}

	w->setName(owner.getProperty (Ids::name).toString() + "/" + w->getContent()->getContentName());
	windows.add (w);

	return (w);
}
```

**Comment by damiensellier** [comment]
CtrlrManagerInstance.cpp

```
    if (getInstanceMode() == InstanceSingle || getInstanceMode() == InstanceSingleRestriced) // Added v5.6.31. Force LnF V3 for Main & Child Windows if panel is Legacy
    {
        ValueTree ed = getInstanceTree().getChildWithName(Ids::uiPanelEditor);
        
        if (ed.getProperty(Ids::uiPanelLegacyMode) == "1"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V3"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V2"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V1")
        {
            setProperty(Ids::ctrlrLookAndFeel, "V3");
            setProperty(Ids::uiPanelLegacyMode, "1");
        }
        else
        {
            setProperty(Ids::ctrlrLookAndFeel, "V4");
            setProperty(Ids::uiPanelLegacyMode, "0");
        }
    }
```


CtrlrManager.cpp

```

    // setProperty (Ids::ctrlrLookAndFeel, "V4"); // Required upgrade to unify design
    // setProperty (Ids::ctrlrColourScheme, "Light"); // Required upgrade to unify design. See colourSchemeFromProperty()
```

CtrlrEditor.cpp L133

```
    // setColourScheme(gui::colourSchemeFromProperty(owner.getProperty(Ids::ctrlrColourScheme))); // Updated v5.6.31. Sets the LookAndFeel_V4 colourScheme from the Ctrlr General Preferences, not from the loaded Panel
    getLookAndFeel().setUsingNativeAlertWindows((bool)owner.getProperty(Ids::ctrlrNativeAlerts)); // Sets OS Native alert windows or JUCE
```


All combined properly should do the job

**Comment by damiensellier** [comment]
**CtrlrEditor.cpp** L132

FIXED v5.6.31

```
if (owner.getProperty(Ids::uiPanelLegacyMode) == "1") // Added v5.6.31. Force exported instances to get LnF v3 for legacy mode if panel has LnF in legacyMode
    {
        LookAndFeel::setDefaultLookAndFeel(new LookAndFeel_V3()); // Added v5.6.31
        setLookAndFeel(new LookAndFeel_V3()); // Added v5.6.31
    }
    else
    {
        setColourScheme(gui::colourSchemeFromProperty(owner.getProperty(Ids::ctrlrColourScheme))); // Added v5.6.31. Sets the LookAndFeel_V4 colourScheme from the Ctrlr General Preferences, not from the loaded Panel
    }
```

**CtrlrManagerInstance.cpp** L6

```
void CtrlrManager::setEmbeddedDefaults()
{
	_DBG("CtrlrManager::setEmbeddedDefaults");

	/* first init the ApplicationProperties stuff */
	if (ctrlrPlayerInstanceTree.hasProperty(Ids::restricted))
	{
		ctrlrPlayerInstanceMode	= (CtrlrInstance)(int)ctrlrPlayerInstanceTree.getProperty(Ids::restricted);
	}
	else
	{
		ctrlrPlayerInstanceMode = InstanceSingle;
	}

	ctrlrProperties.reset(new CtrlrProperties (*this));

	setProperty (Ids::ctrlrLogToFile, false);
	setProperty (Ids::ctrlrLuaDebug, false);
	setProperty (Ids::ctrlrVersionSeparator, "_");
	setProperty (Ids::ctrlrVersionCompressed, false);
	setProperty (Ids::ctrlrMidiMonInputBufferSize, 8192);
	setProperty (Ids::ctrlrMidiMonOutputBufferSize, 8192);
	setProperty (Ids::ctrlrLuaDisabled, false);
	setProperty (Ids::ctrlrOverwriteResources, true);
	setProperty (Ids::ctrlrAutoSave, true);
	setProperty (Ids::ctrlrAutoSaveInterval, 300);
    // setProperty (Ids::ctrlrLogOptions, 32); // Updated v5.6.31
    setProperty (Ids::ctrlrLogOptions, 6014); // Added v5.6.31. 6014 shows everything by default with MIDI messages in Hex
    setProperty (Ids::ctrlrScrollbarThickness, 16.0f); // Added v5.6.31
    setProperty (Ids::ctrlrMenuBarHeight, 24); // Added v5.6.31
    setProperty (Ids::ctrlrTabBarDepth, 24); // Added v5.6.31
    setProperty (Ids::ctrlrUseEditorWrapper, (bool)ctrlrPlayerInstanceTree.hasProperty(Ids::ctrlrUseEditorWrapper) ? (bool)ctrlrPlayerInstanceTree.getProperty(Ids::ctrlrUseEditorWrapper) : true);
    setProperty (Ids::uiLuaConsoleInputRemoveAfterRun, true);
    
    if (getInstanceMode() == InstanceSingle || getInstanceMode() == InstanceSingleRestriced) // Added v5.6.31. Force LnF V3 for Main & Child Windows if panel is Legacy
    {
        ValueTree ed = getInstanceTree().getChildWithName(Ids::uiPanelEditor);
        
        if (ed.getProperty(Ids::uiPanelLegacyMode) == "1"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V3"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V2"
            || ed.getProperty(Ids::uiPanelLookAndFeel) == "V1")
        {
            setProperty(Ids::uiPanelLegacyMode, "1");
            setProperty(Ids::ctrlrLookAndFeel, "V3");
        }
        else
        {
            setProperty(Ids::uiPanelLegacyMode, "0");
            setProperty(Ids::ctrlrLookAndFeel, "V4");
            
            // Requires passing the colourScheme to the property ctrlrColourScheme from uiPanelLookAndFeel
            if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 Light")
            {
                setProperty(Ids::ctrlrColourScheme, "Light");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 Grey")
            {
                setProperty(Ids::ctrlrColourScheme, "Grey");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 Dark")
            {
                setProperty(Ids::ctrlrColourScheme, "Dark");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 Midnight")
            {
                setProperty(Ids::ctrlrColourScheme, "Midnight");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 JetBlack")
            {
                setProperty(Ids::ctrlrColourScheme, "JetBlack");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 YamDX")
            {
                setProperty(Ids::ctrlrColourScheme, "YamDX");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 AkAPC")
            {
                setProperty(Ids::ctrlrColourScheme, "AkAPC");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 AkMPC")
            {
                setProperty(Ids::ctrlrColourScheme, "AkMPC");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 LexiBlue")
            {
                setProperty(Ids::ctrlrColourScheme, "LexiBlue");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 KurzGreen")
            {
                setProperty(Ids::ctrlrColourScheme, "KurzGreen");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 KorGrey")
            {
                setProperty(Ids::ctrlrColourScheme, "KorGrey");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 KorGold")
            {
                setProperty(Ids::ctrlrColourScheme, "KorGold");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 ArturOrange")
            {
                setProperty(Ids::ctrlrColourScheme, "ArturOrange");
            }
            else if (ed.getProperty(Ids::uiPanelLookAndFeel) == "V4 AiraGreen")
            {
                setProperty(Ids::ctrlrColourScheme, "AiraGreen");
            }
        }
    }
}
```

**Comment by damiensellier** [comment]
![Capture d’écran, le 2024-12-19 à 13 47 00](https://github.com/user-attachments/assets/45fc2c1f-6175-4616-8360-566959af8056)
![Capture d’écran, le 2024-12-19 à 13 47 14](https://github.com/user-attachments/assets/317490ec-27d1-4a28-8b5a-105f2ad97e79)
![Capture d’écran, le 2024-12-19 à 13 47 43](https://github.com/user-attachments/assets/b60d25ec-f8ac-4b6d-b8df-3b69fb5997d5)

**Comment by damiensellier** [comment]
FIXED 20 DEC 2024 for v5.6.31
CtrlrValueTreeEditor.cpp requires defaultLookAndFeel to apply the related colours definitions :


```
// GUI For TreeView Items. Updated v5.6.31
void CtrlrValueTreeEditorItem::paintItem (Graphics &g, int width, int height)
{
    std::unique_ptr<Drawable> icon(provider.getIconForItem (treeToEdit));
	auto &currentLnF = getDefaultLookAndFeel(); // Added v5.6.31
    
    if (isSelected())
	{
        g.setColour(currentLnF.findColour(TextEditor::highlightedTextColourId));
        gui::drawSelectionRectangle (g, width, height, currentLnF.findColour(Slider::rotarySliderFillColourId)); // Updated v5.6.31
        // gui::drawSelectionRectangle (g, width, height)); // Removed v5.6.31
	}
	// g.setColour (Colours::black); // Removed v5.6.31
    g.setColour(currentLnF.findColour(TextEditor::textColourId)); // Added v5.6.31
    
    setDrawsInRightMargin(true);
    
	AttributedString as = provider.getDisplayString(treeToEdit);
	as.setJustification (Justification (Justification::centredLeft));
	as.draw (g, Rectangle <float> (24.0, 0.0, width - 24.0, height));

	if (icon)
		icon->drawWithin(g, Rectangle<float>(4,0, 16, height), RectanglePlacement::centred, 1.0f);
	// g.drawImageWithin (icon, 4, 0, 16, height, RectanglePlacement (RectanglePlacement::centred));
}
```

![Capture d’écran, le 2024-12-20 à 14 19 02](https://github.com/user-attachments/assets/28832144-2b44-47d5-8766-3dd0831fd4d4)

**Comment by damiensellier** [comment]
Min width of background for selected item from 128 to 256px

```
int	CtrlrValueTreeEditorItem::getItemWidth () const
{
	return (jmax<int>(256, 48 + provider.getItemFont(treeToEdit).getStringWidth (provider.getUniqueName(treeToEdit)))); // Updated v5.6.31
}
```
